beacon-chain/sync/validate_data_column.go

```diff
@@ -0,0 +1,291 @@
+	if err := verifier.SidecarKzgProofVerified(); err != nil {
+		return pubsub.ValidationReject, err
+	}
+
+	// [IGNORE] The sidecar is the first sidecar for the tuple `(block_header.slot, block_header.proposer_index, sidecar.index)`
+	// with valid header signature, sidecar inclusion proof, and kzg proof.
+	if s.hasSeenDataColumnIndex(roDataColumn.Slot(), roDataColumn.ProposerIndex(), roDataColumn.DataColumnSidecar.Index) {
+		return pubsub.ValidationIgnore, nil
+	}
+
+	// [REJECT] The sidecar is proposed by the expected `proposer_index` for the block's slot in the context of the current shuffling (defined by block_header.parent_root/block_header.slot).
+	// If the `proposer_index` cannot immediately be verified against the expected shuffling, the sidecar MAY be queued for later processing while proposers for the block's branch are calculated
+	// -- in such a case do not REJECT, instead IGNORE this message.
+	if err := verifier.SidecarProposerExpected(ctx); err != nil {
+		return pubsub.ValidationReject, err
+	}
+
+	// Get the time at slot start.
+	startTime, err := slots.ToTime(uint64(s.cfg.chain.GenesisTime().Unix()), roDataColumn.SignedBlockHeader.Header.Slot)
+	if err != nil {
+		return pubsub.ValidationIgnore, err
+	}
+
+	verifiedRODataColumns, err := verifier.VerifiedRODataColumns()
+	if err != nil {
+		// This should never happen.
+		log.WithError(err).WithFields(logging.DataColumnFields(roDataColumn)).Error("Failed to get verified data columns")
+		return pubsub.ValidationIgnore, err
+	}
+
+	verifiedRODataColumnsCount := len(verifiedRODataColumns)
+
+	if verifiedRODataColumnsCount != 1 {
+		// This should never happen.
+		log.WithField("verifiedRODataColumnsCount", verifiedRODataColumnsCount).Error("Verified data columns count is not 1")
+		return pubsub.ValidationIgnore, errors.New("Wrong number of verified data columns")
+	}
+
+	msg.ValidatorData = verifiedRODataColumns[0]
+	dataColumnSidecarVerificationSuccessesCounter.Inc()
+
+	sinceSlotStartTime := receivedTime.Sub(startTime)
+	validationTime := s.cfg.clock.Now().Sub(receivedTime)
+	dataColumnSidecarVerificationGossipHistogram.Observe(float64(validationTime.Milliseconds()))
+
+	peerGossipScore := s.cfg.p2p.Peers().Scorers().GossipScorer().Score(pid)
+
+	select {
+	case s.dataColumnLogCh <- dataColumnLogEntry{
+		Slot:            roDataColumn.Slot(),
+		ColIdx:          roDataColumn.Index,
+		PropIdx:         roDataColumn.ProposerIndex(),
+		BlockRoot:       roDataColumn.BlockRoot(),
+		ParentRoot:      roDataColumn.ParentRoot(),
+		PeerSuffix:      pid.String()[len(pid.String())-6:],
+		PeerGossipScore: peerGossipScore,
+		validationTime:  validationTime,
+		sinceStartTime:  sinceSlotStartTime,
+	}:
+	default:
+		log.WithField("slot", roDataColumn.Slot()).Warn("Failed to send data column log entry")
+	}
+
+	return pubsub.ValidationAccept, nil
+}
+
+// Returns true if the column with the same slot, proposer index, and column index has been seen before.
```

Comment:
Since for a node that subscribe to all the subnets, this is a hot path
Id consider preallocate 96 bytes. Less GC pressure

Reply:
Fixed in cdd7e2137b68eca9aba7e6587c5d0228537bd16c.